future/atomicTransactionComposer.go

```diff
@@ -212,56 +221,104 @@ func (atc *AtomicTransactionComposer) AddMethodCall(params AddMethodCallParams)
 	for i, arg := range params.Method.Args {
-		if _, ok := abi.TransactionArgTypes[arg.AbiType]; ok {
-			txnAndSigner, ok := params.MethodArgs[i].(TransactionWithSigner)
+		argValue := params.MethodArgs[i]
+
+		if arg.IsTransactionArg() {
+			txnAndSigner, ok := argValue.(TransactionWithSigner)
 			if !ok {
 				return fmt.Errorf("invalid arg type, expected transaction")
 			}
 
-			err := atc.validateTransaction(txnAndSigner.Txn)
+			err := atc.validateTransaction(txnAndSigner.Txn, arg.Type)
 			if err != nil {
 				return err
 			}
 			txsToAdd = append(txsToAdd, txnAndSigner)
 		} else {
-			abiType, err := abi.TypeOf(arg.AbiType)
+			var abiType abi.Type
+			var err error
+
+			if arg.IsReferenceArg() {
+				refArgIndexToBasicArgIndex[len(refArgTypes)] = len(basicArgTypes)
+				refArgValues = append(refArgValues, argValue)
+				refArgTypes = append(refArgTypes, arg.Type)
+
+				// treat the reference as a uint8 for encoding purposes
+				abiType, err = abi.TypeOf("uint8")
+			} else {
+				abiType, err = arg.GetTypeObject()
+			}
 			if err != nil {
 				return err
 			}
 
-			abiArgs = append(abiArgs, params.MethodArgs[i])
-			abiTypes = append(abiTypes, abiType)
+			basicArgValues = append(basicArgValues, argValue)
+			basicArgTypes = append(basicArgTypes, abiType)
 		}
 	}
 
-	// Up to 16 args can be passed to app call. First is reserved for method selector and last is
-	// reserved for remaining args packaged into one tuple.
-	if len(abiArgs) > 14 {
-		tupleTypes := abiTypes[14:]
-		tupleValues := abiArgs[14:]
-		tupleType, err := abi.MakeTupleType(tupleTypes)
+	var foreignAccounts []string
+	var foreignApps []uint64
+	var foreignAssets []uint64
+	refArgsResolved, err := populateMethodCallReferenceArgs(params.Sender.String(), params.AppID, refArgTypes, refArgValues, &foreignAccounts, &foreignApps, &foreignAssets)
+	if err != nil {
+		return err
+	}
+	for i, resolved := range refArgsResolved {
+		basicArgIndex := refArgIndexToBasicArgIndex[i]
+		// use the foreign array index as the encoded argument value
+		basicArgValues[basicArgIndex] = resolved
+	}
+
+	// Up to 16 app arguments can be passed to app call. First is reserved for method selector,
+	// and the rest are for method call arguments. But if more than 15 method call arguments
+	// are present, then the 14th+ are placed in a tuple in the last app argument slot
+	if len(basicArgValues) > 15 {
+		typesForTuple := make([]abi.Type, len(basicArgTypes)-14)
+		copy(typesForTuple, basicArgTypes[14:])
+
+		valueForTuple := make([]interface{}, len(basicArgValues)-14)
+		copy(valueForTuple, basicArgValues[14:])
+
+		tupleType, err := abi.MakeTupleType(typesForTuple)
 		if err != nil {
 			return err
 		}
 
-		abiArgs = abiArgs[:14]
-		abiArgs = append(abiArgs, tupleValues)
-		abiTypes = abiTypes[:14]
-		abiTypes = append(abiTypes, tupleType)
+		basicArgValues = append(basicArgValues[:14], valueForTuple)
+		basicArgTypes = append(basicArgTypes[:14], tupleType)
 	}
```

Comment:
The cucumber tests exposed a bug in this section of code. The problem was this part:

```go
tupleValues := abiArgs[14:]
// ...
abiArgs = abiArgs[:14]
abiArgs = append(abiArgs, tupleValues)
```

which is an issue because `abiArgs[14:]` and `abiArgs[:14]` are backed by the same array, and `append(abiArgs[:14], abiArgs[14:])` does not really append to `abiArgs[:14]`, it sets index 15 to `abiArgs[14:]`, **which contains index 15**. So I believe you end up with the first element of `tupleValues` being a reference to itself.

The solution was to make a copy of the tuple part of the array so that it does not get modified when it's appended back into the original array.

Reply:
I believe goal's implementation suffers from this as well, and Python and Java may also be vulnerable to it, but I haven't checked.

Reply:
oof, it is such a weird thing that the index 15 (count from 0) is a reference to itself after appending, nice catch.
I don't think it happens in Java, I handled it in another way.
I also don't think it will happen in Python, can be verified with some lines of code:
```py
stuff = [i for i in range(20)]
stuff_tail = stuff[14:]
stuff = stuff[:14]
stuff.append(stuff_tail)
print(stuff)
```